DBZ-644 add JDBC data type as additional metadata to DBZ events, if config is opted in

orrganani opened the pull request.

jpechane commented:
@orrganani Looks good, thanks! I'd like just to ask you to add a test to verify that it works.
Also it would be good to look at actual messages delivered to Kafka in both JSON and Avro format to confirm the data are stored there.

orrganani commented:
@jpechane I'm sorry but I accidentally sent you a previous version.
I've updated it now and added a test for the OriginalDataTypeMapper.

Regarding Kafka for both formats - How would I check that?
I've checked it with my application as I am using debezium-embedded in my application and it works just fine

jpechane commented:
Hi Orr, I am sorry I was under other pressing tasks. The basci functionality looks fine. What I need you to do now is to better embed the config option into the framework.

Please look at `io.debezium.connector.mysql.MySqlConnectorConfig`. We need a config field for `column.add.original.type` option. The option then would be retrived via `config.getBoolean()` instead of `forEachMatchingFieldName`.

Also you can test the functionality by extending `MySqlConnectorIT` test. This way we will have almost end-to-end test.

The Kafka testing - this does not need an automated test. It is sufficient if it is done manually. The easiest thing would be to use https://github.com/debezium/debezium-examples/tree/master/tutorial example for both JSON and Avro formats. The only caveat is that you need to build your own Connect image that would contain the modified connector. If you will struggle with this part then I can make it on my own.

jpechane commented:
@orrganani Hi, do you plan to continue on this one?

orrganani commented:
Hi @jpechane, sorry for the unavailability, we've been having some production issues lately.
I do plan on continuing this, early next week.
Is it delaying the next release?

jpechane commented:
@orrganani Hi, no worries take your time, I was just checking the status

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/Filters.java:
```diff
@@ -99,6 +99,7 @@ public Filters(Configuration config) {
         ColumnMappers.Builder columnMapperBuilder = ColumnMappers.create();
         config.forEachMatchingFieldNameWithInteger("column\\.truncate\\.to\\.(\\d+)\\.chars", columnMapperBuilder::truncateStrings);
         config.forEachMatchingFieldNameWithInteger("column\\.mask\\.with\\.(\\d+)\\.chars", columnMapperBuilder::maskStrings);
+        config.forEachMatchingFieldName("column\\.add\\.original\\.type", columnMapperBuilder::addOriginalType);
```
Nice idea with using the column mappers!

gunnarmorling commented on debezium-core/src/main/java/io/debezium/relational/mapping/AddOriginalDataType.java:
```diff
@@ -0,0 +1,23 @@
+import io.debezium.relational.ValueConverter;
+import org.apache.kafka.connect.data.SchemaBuilder;
+
+public class AddOriginalDataType implements ColumnMapper {
```
Can you add a class comment describing its purpose?

gunnarmorling commented on debezium-core/src/main/java/io/debezium/relational/mapping/AddOriginalDataType.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public void alterFieldSchema(Column column, SchemaBuilder schemaBuilder) {
+       schemaBuilder.parameter("originalType", String.valueOf(column.jdbcType()));
+       schemaBuilder.parameter("columnSize", String.valueOf(column.length()));
```
I think I'd prefer a single parameter, "originalType", comprising type name, length and scale (i.e. like the first part of the `toString()` implementation of `ColumnImpl`.

gunnarmorling commented on debezium-core/src/test/java/io/debezium/relational/mapping/OriginalDataTypeTest.java:
```diff
@@ -0,0 +1,37 @@
+        .create();
+
+    @Test
+    public void shouldTruncateStrings() {
```
The name seems a bit odd?

Can you also add the `FixFor` annotation, referencing the issue's JIRA number?

orrganani commented:
@gunnarmorling  I'm on vacation until mid next week, so I asked Sapir from my team to pick up the pull request.

gunnarmorling commented:
Ok, thanks for the update. No worries, though. If it misses this week's release, the next one after that will be in three weeks.

gunnarmorling commented:
Hey @orrganani, just checking in on this one. Any plans on your side for finishing it up? Thanks!

gunnarmorling commented:
Hey @orrganani, this looks almost done. So I think I'll take over and complete it. Once it's merged you can take a look and let us know if you see anything else to be done.

orrganani commented:
guys I'm terribly sorry that this is taking so long from my side we just don't have the capacity currently to finish up the PR. 

if you would like to finish it up that would be great, thanks 
I'll take a look after it's merge in

gunnarmorling commented:
@orrganani No worries, we've all been there :) I'll finish it up then, shouldn't take long.

My only question to you is: will it suffice to add the source column type's name as a parameter (e.g. `__debezium.source.column.type=MEDIUMINT`? From what I've read in the JIRA issue, it should, as you can get the precision/scale via existing parameters already for NUMERIC/DECIMAL columns (where it's variable). I don't like adding them for other (non-variable) types (e.g. MEDIUMINT) too much, as it either may cause trouble when re-using the already used keys or it would be redundant when using other keys in some cases then. But from the type name you should be able to derive the correct width for the downstream DB.
